cglib/src/test/java/net/sf/cglib/classloader/FilteringClassLoader.java

```diff
@@ -0,0 +1,119 @@
+
+    public FilteringClassLoader(ClassLoader parent, ClassLoader delegatingClassLoader, String[] filterPrefixIncludes, String[] filterPrefixExcludes) {
+        super(parent);
+        this.delegatingClassLoader = delegatingClassLoader;
+        this.filterPrefixIncludes = filterPrefixIncludes;
+        this.filterPrefixExcludes = filterPrefixExcludes;
+    }
+
+    public FilteringClassLoader(ClassLoader parent, ClassLoader delegatingClassLoader, String ... filterPrefixIncludes) {
+        this(parent, delegatingClassLoader, filterPrefixIncludes, null);
+    }
+
+    public Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
+        if (needFiltering(name)) {
+            synchronized(this) {
+                // Check whether this class is already loaded by this classloader
+                Class<?> cl = findLoadedClass(name);
+                // Do not load from parent classloader - try to find class file in delegating classloader and load it in this classloader
+                if (cl == null && delegatingClassLoader != null) {
+                    InputStream inputStream = delegatingClassLoader.getResourceAsStream(name.replace('.','/')+".class");
+                    if (inputStream != null) {
+                        // Class file was found - load it
+                        ByteArrayOutputStream targetBuffer = new ByteArrayOutputStream();
+                        try {
+                            try {
+                                byte[] buffer = new byte[1024];
+                                for(int r = inputStream.read(buffer, 0, buffer.length); r > 0; r = inputStream.read(buffer, 0, buffer.length)) {
+                                    targetBuffer.write(buffer, 0, r);
+                                }
+                            } finally {
+                                inputStream.close();
+                            }
+                        } catch (IOException e) {
+                            throw new ClassNotFoundException("Cannot load class with name '"+name+"' from delegating classloader", e);
+                        }
+                        byte[] classBytes = targetBuffer.toByteArray();
+                        // Define and resolve class (if required)
+                        cl = defineClass(name, classBytes, 0, classBytes.length);
+                        if (resolve) {
+                            resolveClass(cl);
+                        }
+                    }
+                }
+                // Filtering result - either class loaded in this classloader or exception
+                if (cl != null) {
+                    return cl;
+                } else {
+                    throw new ClassNotFoundException("Class with name '"+name+"' should not be loaded from parent classloader and is not found in delegating classloader");
+                }
+            }
+        } else {
+            return super.loadClass(name, resolve);
```

Comment:
It makes sense to reorder if branch to avoid `if (...) { long code ... long code} else {return}` pattern